settlement/dymension/dymension.go

```diff
@@ -215,76 +216,96 @@ func (d *HubClient) PostBatch(batch *types.Batch, daClient da.Client, daResult *
+						"startHeight",
+						batch.StartHeight,
+						"endHeight",
+						batch.EndHeight,
+						"error",
+						err,
+					)
+					// If submission (i.e. only submission, not acceptance) fails we try submitting again.
+					// Sleep to allow context cancellation to take effect before retrying
+					time.Sleep(100 * time.Millisecond)
+					continue
+				}
 			}
+			break
 		}
 
 		// Batch was submitted successfully. Wait for it to be accepted by the settlement layer.
 		timer := time.NewTimer(d.batchAcceptanceTimeout)
 		defer timer.Stop()
-
-		select {
-		case <-d.ctx.Done():
-			return d.ctx.Err()
-
-		case <-subscription.Cancelled():
-			return fmt.Errorf("subscription cancelled")
-
-		case <-subscription.Out():
-			d.logger.Debug("Batch accepted", "startHeight", batch.StartHeight, "endHeight", batch.EndHeight)
-			return nil
-
-		case <-timer.C:
-			// Check if the batch was accepted by the settlement layer, and we've just missed the event.
-			includedBatch, err := d.waitForBatchInclusion(batch.StartHeight)
-			if err != nil {
-				d.logger.Error(
-					"Wait for batch inclusion",
-					"startHeight",
-					batch.StartHeight,
-					"endHeight",
-					batch.EndHeight,
-					"error",
-					err,
-				)
-
-				timer.Stop() // we don't forget to clean up
-				continue
+		batchAcceptanceAttempts := d.batchAcceptanceAttempts
+
+		for {
+			select {
+			case <-d.ctx.Done():
+				return d.ctx.Err()
+
+			case <-subscription.Cancelled():
+				return fmt.Errorf("subscription cancelled")
+
+			case event := <-subscription.Out():
+				eventData, _ := event.Data().(*settlement.EventDataNewBatchAccepted)
+				if eventData.EndHeight != batch.EndHeight {
+					d.logger.Info("Received event for a different batch, ignoring.", "event", eventData)
+					continue
+				}
+				d.logger.Debug("Batch accepted.", "startHeight", batch.StartHeight, "endHeight", batch.EndHeight)
+				return nil
+
+			case <-timer.C:
+				// Check if the batch was accepted by the settlement layer, and we've just missed the event.
+				includedBatch, err := d.pollForBatchInclusion(batch.EndHeight)
+				if err == nil && !includedBatch {
+					// no error, but still not included
+					batchAcceptanceAttempts--
+					if batchAcceptanceAttempts > 0 {
+						timer.Reset(d.batchAcceptanceTimeout)
+						continue
+					}
+					err = fmt.Errorf("timed out waiting for batch inclusion on settlement layer")
+				}
+
+				if err != nil {
+					d.logger.Error(
+						"Wait for batch inclusion",
+						"startHeight",
+						batch.StartHeight,
+						"endHeight",
+						batch.EndHeight,
+						"error",
+						err,
+					)
+					// If errored polling, start again the submission loop.
+					break
+				}
+
+				// all good
+				d.logger.Info("Batch accepted", "startHeight", batch.StartHeight, "endHeight", batch.EndHeight)
+				return nil
 			}
-
-			// all good
-			d.logger.Info("Batch accepted", "startHeight", includedBatch.StartHeight, "endHeight", includedBatch.EndHeight)
-			return nil
 		}
 	}
 }
```

Comment:
this whole block is getting pretty complicated, can we extract some methods (`submitBatch`, `waitForInclusion`) and try to simplify (especially to simplify the submission part)

Reply:
simplified a bit